Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/SubmissionTask.kt

```diff
@@ -0,0 +1,114 @@
+    private val exposureKeyHistoryCalculations: ExposureKeyHistoryCalculations
+) : Task<DefaultProgress, Task.Result> {
+
+    private val internalProgress = ConflatedBroadcastChannel<DefaultProgress>()
+    override val progress: Flow<DefaultProgress> = internalProgress.asFlow()
+
+    private var isCanceled = false
+
+    override suspend fun run(arguments: Task.Arguments) = try {
+        Timber.d("Running with arguments=%s", arguments)
+        with(arguments as Arguments) {
+            playbook.submission(
+                Playbook.SubmissionData(
+                    registrationToken,
+                    getHistory(),
+                    true,
+                    supportedCountries()
+                ).also {
+                    checkCancel()
+                }
+            )
+        }
+        SubmissionService.submissionSuccessful()
+        object : Task.Result {}
+    } catch (error: Exception) {
+        Timber.tag(TAG).e(error)
+        throw error
+    } finally {
+        Timber.i("Finished (isCanceled=$isCanceled).")
+        internalProgress.close()
+    }
+
+    private fun Arguments.getHistory(): List<KeyExportFormat.TemporaryExposureKey> =
+        exposureKeyHistoryCalculations.transformToKeyHistoryInExternalFormat(
+            keys,
+            symptoms
+        )
+
+    private suspend fun supportedCountries(): List<String> {
+        appConfigProvider.getAppConfig().apply {
+            return if (supportedCountriesList.isEmpty()) {
+                toNewConfig {
+                    addSupportedCountries(FALLBACK_COUNTRY)
+                }.also {
+                    Timber.w("Country list was empty, corrected: %s", it.supportedCountriesList)
+                }
+            } else {
+                this
+            }.supportedCountriesList
+        }
+    }
+
+    private fun checkCancel() {
+        if (isCanceled) throw TaskCancellationException()
+    }
+
+    override suspend fun cancel() {
+        Timber.w("cancel() called.")
+        isCanceled = true
+    }
+
+    class Arguments(
+        val registrationToken: String,
+        val keys: List<TemporaryExposureKey>,
+        val symptoms: Symptoms
+    ) : Task.Arguments
+
+    data class Config(
+        override val executionTimeout: Duration = SUBMISSION_TASK_TIMEOUT,//TODO unit-test that not > 9 min
+
+        override val collisionBehavior: TaskFactory.Config.CollisionBehavior =
+            TaskFactory.Config.CollisionBehavior.ENQUEUE
+
+    ) : TaskFactory.Config
+
+    class Factory @Inject constructor(
+        private val taskByDagger: Provider<SubmissionTask>
+    ) : TaskFactory<DefaultProgress, Task.Result> {
+
+        override val config: TaskFactory.Config = Config()
+        override val taskProvider: () -> Task<DefaultProgress, Task.Result> = {
+            taskByDagger.get()
+        }
+    }
+
+    companion object {
+        private val SUBMISSION_TASK_TIMEOUT = Duration.standardMinutes(8)
```

Comment:
Can we inline this?

Reply:
I agree, its only used in one place and the variable that its assigned to is already descriptive enough in my opinion